File: backend/pkg/models/assignment.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// =============================================================================
// Subdocuments
// =============================================================================

type AssignmentShift struct {
	StartTime string `json:"start_time" bson:"start_time"` // "HH:mm"
	EndTime   string `json:"end_time" bson:"end_time"`     // "HH:mm"
}

type AssignmentOriginalShift struct {
	Type      string `json:"type" bson:"type"`                     // "shift" | "off"
	StartTime string `json:"start_time,omitempty" bson:"start_time,omitempty"` // "HH:mm" if type="shift"
	EndTime   string `json:"end_time,omitempty" bson:"end_time,omitempty"`     // "HH:mm" if type="shift"
	Source    string `json:"source,omitempty" bson:"source,omitempty"`         // optional: "working_hours"|"manual"|"schedule"
}

type AssignmentDayOffReward struct {
	Eligible           bool       `json:"eligible" bson:"eligible"`
	Status             string     `json:"status" bson:"status"` // none|pending|granted|used|cancelled
	Description        string     `json:"description,omitempty" bson:"description,omitempty"`
	GrantedAt          *time.Time `json:"granted_at,omitempty" bson:"granted_at,omitempty"`
	UsedAt             *time.Time `json:"used_at,omitempty" bson:"used_at,omitempty"`
	ReplacementOffDate *time.Time `json:"replacement_off_date,omitempty" bson:"replacement_off_date,omitempty"`
}

type AssignmentEmployee struct {
	UserID         primitive.ObjectID      `json:"user_id" bson:"user_id"`
	OriginalShift  AssignmentOriginalShift  `json:"original_shift" bson:"original_shift"`
	AssignedShift  AssignmentShift          `json:"assigned_shift" bson:"assigned_shift"`
	EmployeeStatus string                  `json:"employee_status" bson:"employee_status"` // pending|agreed|rejected|cancelled
	RejectionNote  string                  `json:"rejection_note,omitempty" bson:"rejection_note,omitempty"`
	ConfirmedAt    *time.Time              `json:"confirmed_at,omitempty" bson:"confirmed_at,omitempty"`
	DayOffReward   AssignmentDayOffReward  `json:"day_off_reward" bson:"day_off_reward"`
}

// =============================================================================
// Main document
// =============================================================================

type Assignment struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	DepartmentID  primitive.ObjectID `json:"department_id" bson:"department_id"`
	RequestedByID primitive.ObjectID `json:"requested_by_id" bson:"requested_by_id"`

	Date   time.Time `json:"date" bson:"date"`     // only date
	Reason string    `json:"reason" bson:"reason"` // reason/event/occupancy

	Status string `json:"status" bson:"status"` // draft|submitted|published|cancelled
	Notes  string `json:"notes,omitempty" bson:"notes,omitempty"`

	ShiftTarget AssignmentShift `json:"shift_target" bson:"shift_target"`

	Employees []AssignmentEmployee `json:"employees" bson:"employees"`

	CreatedAt time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt time.Time `json:"updated_at" bson:"updated_at"`
}

// =============================================================================
// Constants
// =============================================================================

const (
	AssignmentStatusDraft     = "draft"
	AssignmentStatusSubmitted = "submitted"
	AssignmentStatusPublished = "published"
	AssignmentStatusCancelled = "cancelled"
)

const (
	ShiftTypeShift = "shift"
	ShiftTypeOff   = "off"
)

const (
	AssignmentEmployeeStatusPending   = "pending"
	AssignmentEmployeeStatusAgreed    = "agreed"
	AssignmentEmployeeStatusRejected  = "rejected"
	AssignmentEmployeeStatusCancelled = "cancelled"
)

const (
	DayOffRewardStatusNone      = "none"
	DayOffRewardStatusPending   = "pending"
	DayOffRewardStatusGranted   = "granted"
	DayOffRewardStatusUsed      = "used"
	DayOffRewardStatusCancelled = "cancelled"
)

// =============================================================================
// Request Models
// =============================================================================

type AssignmentEmployeeInput struct {
	UserID string `json:"user_id" binding:"required"`

	// optional: jika Anda ingin override shift per karyawan
	AssignedStartTime *string `json:"assigned_start_time,omitempty"` // "HH:mm"
	AssignedEndTime   *string `json:"assigned_end_time,omitempty"`   // "HH:mm"
}

type CreateAssignmentRequest struct {
	DepartmentID  string                  `json:"department_id" binding:"required"`
	Date          string                  `json:"date" binding:"required"` // YYYY-MM-DD
	ShiftStart    string                  `json:"start_time" binding:"required"`
	ShiftEnd      string                  `json:"end_time" binding:"required"`
	Reason        string                  `json:"reason" binding:"required"`
	Status        string                  `json:"status,omitempty"` // draft/submitted
	Employees     []AssignmentEmployeeInput `json:"employees" binding:"required,min=1"`
	Notes         string                  `json:"notes,omitempty"`
}

type UpdateAssignmentRequest struct {
	Date       *string `json:"date,omitempty"` // YYYY-MM-DD
	ShiftStart *string `json:"start_time,omitempty"`
	ShiftEnd   *string `json:"end_time,omitempty"`
	Reason     *string `json:"reason,omitempty"`
	Status     *string `json:"status,omitempty"`
	Notes      *string `json:"notes,omitempty"`

	Employees *[]AssignmentEmployeeInput `json:"employees,omitempty"`
}

// karyawan setuju / tolak
type UpdateAssignmentEmployeeStatusRequest struct {
	Status        string `json:"status" binding:"required"` // agreed|rejected
	RejectionNote string `json:"rejection_note,omitempty"`
}

// optional: ketika karyawan/HR menetapkan tanggal off pengganti (jika dipakai)
type UseReplacementDayOffRequest struct {
	ReplacementOffDate string `json:"replacement_off_date" binding:"required"` // YYYY-MM-DD
}

// =============================================================================
// Response Models
// =============================================================================

type AssignmentEmployeeResponse struct {
	UserID         string     `json:"user_id"`
	FullName       string     `json:"full_name"`
	PayrollNumber  string     `json:"payroll_number"`
	PositionName   string     `json:"position_name"`

	OriginalShiftType string `json:"original_shift_type"`
	OriginalStartTime string `json:"original_start_time,omitempty"`
	OriginalEndTime   string `json:"original_end_time,omitempty"`

	AssignedStartTime string `json:"assigned_start_time"`
	AssignedEndTime   string `json:"assigned_end_time"`

	EmployeeStatus string     `json:"employee_status"`
	RejectionNote  string     `json:"rejection_note,omitempty"`
	ConfirmedAt    *time.Time `json:"confirmed_at,omitempty"`

	DayOffEligible           bool       `json:"day_off_eligible"`
	DayOffStatus             string     `json:"day_off_status"`
	DayOffGrantedAt          *time.Time `json:"day_off_granted_at,omitempty"`
	DayOffUsedAt             *time.Time `json:"day_off_used_at,omitempty"`
	ReplacementOffDate       *time.Time `json:"replacement_off_date,omitempty"`
}

type AssignmentResponse struct {
	ID             string    `json:"id"`
	DepartmentID   string    `json:"department_id"`
	DepartmentName string    `json:"department_name"`

	RequestedByID   string `json:"requested_by_id"`
	RequestedByName string `json:"requested_by_name"`

	Date   time.Time `json:"date"`
	Reason string    `json:"reason"`

	Status string `json:"status"`
	Notes  string `json:"notes,omitempty"`

	ShiftStart string `json:"start_time"`
	ShiftEnd   string `json:"end_time"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	Employees []AssignmentEmployeeResponse `json:"employees"`
}

File: mobile_app/lib/models/assignment.dart
// lib/models/assignment.dart

class Assignment {
  final String id;
  final String departmentId;
  final String departmentName;
  final String requestedById;
  final String requestedByName;
  final DateTime date;
  final String reason;
  final String status;
  final String? notes;
  final String startTime;
  final String endTime;
  final List<AssignmentEmployee> employees;
  final DateTime createdAt;
  final DateTime updatedAt;

  Assignment({
    required this.id,
    required this.departmentId,
    required this.departmentName,
    required this.requestedById,
    required this.requestedByName,
    required this.date,
    required this.reason,
    required this.status,
    this.notes,
    required this.startTime,
    required this.endTime,
    required this.employees,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Assignment.fromJson(Map<String, dynamic> json) {
    DateTime parseDate(dynamic v) {
      if (v == null) return DateTime.now();
      try {
        return DateTime.parse(v.toString()).toLocal();
      } catch (_) {
        return DateTime.now();
      }
    }

    final emps = (json['employees'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(AssignmentEmployee.fromJson)
            .toList() ??
        [];

    return Assignment(
      id: (json['id'] ?? '').toString(),
      departmentId: (json['department_id'] ?? '').toString(),
      departmentName: (json['department_name'] ?? json['departmentName'] ?? '').toString(),
      requestedById: (json['requested_by_id'] ?? '').toString(),
      requestedByName: (json['requested_by_name'] ?? json['requestedByName'] ?? '').toString(),
      date: parseDate(json['date']),
      reason: (json['reason'] ?? '').toString(),
      status: (json['status'] ?? 'draft').toString().toLowerCase(),
      notes: json['notes']?.toString(),
      startTime: (json['start_time'] ?? json['shift_start'] ?? json['startTime'] ?? '').toString(),
      endTime: (json['end_time'] ?? json['shift_end'] ?? json['endTime'] ?? '').toString(),
      employees: emps,
      createdAt: parseDate(json['created_at'] ?? json['createdAt']),
      updatedAt: parseDate(json['updated_at'] ?? json['updatedAt']),
    );
  }

  String get statusDisplay {
    switch (status) {
      case 'draft':
        return 'Draft';
      case 'submitted':
        return 'Dikirim';
      case 'published':
        return 'Dipublikasikan';
      case 'cancelled':
        return 'Dibatalkan';
      default:
        return status;
    }
  }

  bool get isDraft => status == 'draft';
  bool get isSubmitted => status == 'submitted';
  bool get isPublished => status == 'published';
}

class AssignmentEmployee {
  final String userId;
  final String fullName;
  final String payrollNumber;
  final String positionName;
  final String originalStartTime;
  final String originalEndTime;
  final String assignedStartTime;
  final String assignedEndTime;
  final String employeeStatus;
  final String? rejectionNote;
  final DateTime? confirmedAt;
  final bool dayOffEligible;
  final String dayOffStatus;
  final DateTime? dayOffGrantedAt;
  final DateTime? dayOffUsedAt;
  final DateTime? replacementOffDate;

  AssignmentEmployee({
    required this.userId,
    required this.fullName,
    required this.payrollNumber,
    required this.positionName,
    required this.originalStartTime,
    required this.originalEndTime,
    required this.assignedStartTime,
    required this.assignedEndTime,
    required this.employeeStatus,
    this.rejectionNote,
    this.confirmedAt,
    required this.dayOffEligible,
    required this.dayOffStatus,
    this.dayOffGrantedAt,
    this.dayOffUsedAt,
    this.replacementOffDate,
  });

  factory AssignmentEmployee.fromJson(Map<String, dynamic> json) {
    DateTime? parseMaybeDate(dynamic v) {
      if (v == null) return null;
      try {
        return DateTime.parse(v.toString()).toLocal();
      } catch (_) {
        return null;
      }
    }

    return AssignmentEmployee(
      userId: (json['user_id'] ?? json['userId'] ?? '').toString(),
      fullName: (json['full_name'] ?? json['fullName'] ?? json['name'] ?? '').toString(),
      payrollNumber: (json['payroll_number'] ?? '').toString(),
      positionName: (json['position_name'] ?? '').toString(),
      originalStartTime: (json['original_start_time'] ?? '').toString(),
      originalEndTime: (json['original_end_time'] ?? '').toString(),
      assignedStartTime: (json['assigned_start_time'] ?? json['assignedStartTime'] ?? '').toString(),
      assignedEndTime: (json['assigned_end_time'] ?? json['assignedEndTime'] ?? '').toString(),
      employeeStatus: (json['employee_status'] ?? 'pending').toString().toLowerCase(),
      rejectionNote: json['rejection_note']?.toString(),
      confirmedAt: parseMaybeDate(json['confirmed_at'] ?? json['confirmedAt']),
      dayOffEligible: (json['day_off_eligible'] == true) || (json['dayOffEligible'] == true),
      dayOffStatus: (json['day_off_status'] ?? json['dayOffStatus'] ?? '').toString(),
      dayOffGrantedAt: parseMaybeDate(json['day_off_granted_at'] ?? json['dayOffGrantedAt']),
      dayOffUsedAt: parseMaybeDate(json['day_off_used_at'] ?? json['dayOffUsedAt']),
      replacementOffDate: parseMaybeDate(json['replacement_off_date'] ?? json['replacementOffDate']),
    );
  }
}
